src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java

```diff
@@ -249,16 +266,27 @@ public void leaveToken(DetailAST ast) {
     /**
      * Checks whether a class should be declared as final or not.
      *
-     * @param desc description of the class
+     * @param classDesc description of the class
      * @return true if given class should be declared as final otherwise false
      */
-    private static boolean shouldBeDeclaredAsFinal(ClassDesc desc) {
-        return desc.isWithPrivateCtor()
-                && !(desc.isDeclaredAsAbstract()
-                    || desc.isSuperClassOfAnonymousInnerClass())
-                && !desc.isDeclaredAsFinal()
-                && !desc.isWithNonPrivateCtor()
-                && !desc.isWithNestedSubclass();
+    private static boolean shouldBeDeclaredAsFinal(ClassDesc classDesc) {
+        final boolean shouldBeFinal;
+
+        if (classDesc.isWithImplicitConstructor()) {
```

Comment:
Move out common checks and test them first and set `shouldBeFinal` to false if they are set.
Example:
````
if (A) {
  shouldBeFinal = false;
} else if (classDesc.isWithImplicitConstructor()) {
  ...
} else {
  ...
}
````

These are all common:
````
            shouldBeFinal = !(classDesc.isDeclaredAsAbstract() // 2
                    || classDesc.isSuperClassOfAnonymousInnerClass()) // 3
                    && !classDesc.isDeclaredAsFinal() // 1
                    && !classDesc.isWithNestedSubclass() // 4
````

I put comments with suggested order if possible.

Reply:
Done